### PR TITLE
add support for querying by geometry

### PR DESCRIPTION
**Pull request by keewis**
No H3 support, and uses `shapely.Polygon` which might not actually be suitable for this.

### CODE REVIEW

File: xdggs/accessor.py [hunk 1]
```diff
@@ -53,6 +55,12 @@ def sel_latlon(
         cell_indexers = {self._name: self.index._latlon2cellid(latitude, longitude)}
         return self._obj.sel(cell_indexers)
 
+    def query(self, geom: shapely.Geometry, **options) -> xr.Dataset | xr.DataArray:
+        cell_ids = self._index._geom2cellid(geom, options)
+        mask = np.isin(cell_ids, self._index._pd_index.index.values)
```
Comment (benbovy) [inline xdggs/accessor.py:60]: nit: perhaps useful to leave a small comment about why we need masking.

**Review by benbovy — COMMENTED** [review]
Thanks @keewis!

A couple of comments:

- Do we want to support any `shapely.Geometry`? This might be quite ambitious regarding (1) support across multiple grid systems (healpy, h3, etc.) and (2) how we actually select grid cells (intersects vs. within the geometry feature).
- If we only support polygons, perhaps easier would be to just pass the coordinates of the polygon vertices and avoid depending on shapely?
- About the name of the methods, what about this pattern?
    - `Dataset.dggs.sel_from_points(latitude, longitude)`
    - `Dataset.dggs.sel_from_geom(geom)`
    - `Dataset.dggs.sel_from_polygon(interleaved_latlon)` or `Dataset.dggs.sel_from_polygon(latitude, longitude)`

**Comment by benbovy** [comment]
> About the name of the methods, what about this pattern? 

I changed my mind in #20 :). `query` is consistent with Xvec.

However, I wonder if we shouldn't focus first on methods for extracting cell geometry (#10) or for converting DGGS data cubes into vector data cubes (using xvec). Xvec already implements querying by geometry in a very generic way.

There may be more optimal ways to perform spatial queries on DGGS data than considering it just like any arbitrary vector data? From what I've seen, it looks like DGGS tools are doing things internally that are quite similar to what Xvec does (via shapely / GEOS's rtree).

**Comment by keewis** [comment]
well, if that's the case then we should probably make use of `xvec` instead. However, that would mean that we'd have to have `query` add the `boundaries` coordinate and index it if it doesn't exist already (or error, but in that case we might as well not provide it)

**Comment by benbovy** [comment]
> However, that would mean that we'd have to have query add the boundaries coordinate and index it if it doesn't exist already (or error, but in that case we might as well not provide it)

Or do not support any spatial indexing in DGGS for now (no `Dataset.dggs.query()`) and recommend users to do something like below?

```python
# convert to vector data cube
# (automatically set a xvec.GeometryIndex so it is ready for use)
vds = ds.dggs.to_polygons()

# spatial query
vds.xvec.query(...)
```

That doesn't look like much extra work on the user side.

**Comment by keewis** [comment]
okay, I agree with not doing this for now until we have evidence that this is possible for every DGGS using just the cell id

**Comment by tinaok** [comment]
I do want to select the interest area in polygon for my use-case.

**Comment by benbovy** [comment]
> I do want to select the interest area in polygon for my use-case.

Yes, the idea is to support that via Xvec (after converting the DGGS data cube to a polygon vector data cube) and see later if there is a need and/or a more optimal way to support that in xdggs.

**Comment by keewis** [comment]
it does sound like we're not going to pursue this way of implementing the geometry search any further, so let's close this?

**Comment by benbovy** [comment]
Yes we can close this I think, and maybe re-open it later if needed.

**Pull request closed without merging**